### PR TITLE
fix(activity): improve failed MMPay steps in Activity details

### PR DESCRIPTION
**Pull request by vinnyhoward**
## **Description**

A failed MetaMask Pay step looked identical to every other step — same small dot — so nothing said which leg broke, and the reason was unreachable. Failed steps now show a **red cross on the step that actually failed** and open a **bottom sheet explaining what happened**, with the block explorer as a labelled button inside it. The explanation is derived from transaction state rather than error text.

**What**

- **Red cross replaces the dot** on the failed step; steps after it render as upcoming. Non-failed steps are untouched.
- **The failed row opens a sheet instead of the explorer.** One row, one action — the explorer moves into the sheet as a button that links the failed leg's own transaction.
- **New `usePayFundsFailure`** classifies the failure into one of five outcomes from the legs and bridge history, and places the cross on the step its sentence names.
- Covers **Perps and Predict deposits**; withdrawals keep their previous behavior (the copy is deposit-worded).

**Why**

- Error strings are the wrong source. The most common failure — a bridge settling off-chain — records none at all.
- When one does exist it is developer text: `execution reverted`, `replacement transaction underpriced`.
- Two tap targets on one row is not discoverable. Users cannot know a 16px cross does something different from the row around it.

### How the reason is derived

No error-string parsing. Each bucket is read off state, checked earliest-leg-first so the most specific explanation wins, and the cross lands on the step the bucket names.

| Bucket | Detected from | Cross lands on | What the sheet says | Copy |
| ------ | ------------- | -------------- | ------------------- | ---- |
| `approvalFailed` | approve leg status is failed | Approve funds | Nothing left your wallet, safe to retry | placeholder |
| `bridgeFailed` | bridge leg failed, **or** bridge history says `FAILED` | Bridge funds | Funds stayed on the network you paid from | placeholder |
| `bridgedNotDeposited` | bridge history says `COMPLETE` **and** landed on the surface's destination asset | Add funds | Your $12.53 of USDC is on Arbitrum, retry the deposit | **existing copy** |
| `notSubmitted` | parent has no hash | first step | Never sent, nothing left your wallet | placeholder |
| `cancelled` | row status is cancelled | first step | You canceled this deposit | placeholder |
| `unknown` | none of the above matched | last step | Generic | placeholder |

`bridgedNotDeposited` reuses the sentence the pre-redesign transaction details screen ships, parameterised on destination — Perps says USDC on Arbitrum, Predict says USDC.e on Polygon — with the amount from Pay's `targetFiat` (what actually bridged, not the fee-inclusive total). The old screen now consumes the same locale key, so the two cannot drift.

<details>
<summary>More info: copy that still needs a product pass</summary>

Five of the six sentences are the author's and should not ship unreviewed. Under `activity_details.failure`:

- **bridge_failed** — "The bridge didn't complete, so your funds stayed on the network you paid from. Nothing was deposited."
- **approval_failed** — "The approval didn't go through, so nothing left your wallet. Try again."
- **not_submitted** — "This deposit was never sent, so nothing left your wallet."
- **cancelled** — "You canceled this deposit, so nothing was deposited."
- **unknown** — "We couldn't determine why this step failed."

They follow the tone of the one real sentence: what happened, where the money is, what to do next. The ask for product is six outcomes, not error strings.

`bridged_not_deposited` is the existing wording, parameterised on network/symbol; the perps-specific `transaction_details.perps_deposit_solution` key is retired and the pre-redesign screen consumes the shared key.

</details>

<details>
<summary>More info: known gaps and touches outside ActivityDetails</summary>

**Completed-step timestamps are the row's single timestamp**, not per-leg times — pre-existing presentation, unchanged here.

**Non-failed steps still link the parent transaction** on tap; only the failed step's sheet links its culprit leg.

**Withdrawal failure copy does not exist yet.** Withdrawals skip classification entirely, so a failed withdrawal step keeps the old explorer-press behavior.

**Not covered:** the Money tab's transaction details — a separate component owned by confirmations that still shows a red dot. `StepFailureIcon` lives in the shared `StepTimeline` folder so it can adopt this without a copy.

**Touches confirmations (reviewers from that team, please glance):** `transaction-details-status.tsx` now consumes the shared `bridged_not_deposited` locale key, and `useActivityPayMetadata` reads the live transaction rather than the navigation-time snapshot, so the fee section and the failure sheet cannot disagree.

</details>

<details>
<summary>More info: test evidence</summary>

`yarn jest app/components/Views/ActivityDetails app/components/UI/StepTimeline app/components/Views/confirmations/components/activity/transaction-details-status app/components/Views/confirmations/components/progress-list`

```
Test Suites: 43 passed, 43 total
Tests:       378 passed, 378 total
```

New coverage:

- `usePayFundsFailure.test.ts` — all six buckets, approval winning over later legs, off-chain bridge failure, the destination fail-closed guard (COMPLETE on the wrong asset does not claim the destination), classification from bridge history when the leg meta was pruned, $0/malformed amounts falling back to no-amount wording, per-surface destination wording, skip for withdrawals, the failed leg's explorer target.
- `ActivityDetailsStepTimeline.test.tsx` — the sheet opens instead of the explorer, dismisses itself before opening the explorer, shows the updated message when classification changes while open, links the failed leg's target over the row's, failure marked at any step position.
- `ActivityDetailsPerpsStepTimeline.test.tsx` — the cross lands on the step each shape names (with no fabricated timestamps before it), deposit and withdrawal defaults, failed withdrawal steps keep the explorer press.
- `PredictDetails.utils.test.ts` — bridge failures land on the bridge step; deposit failures on the terminal step.

`yarn lint` clean on all touched files. `yarn lint:tsc` reports no new errors.

</details>

## **Changelog**

CHANGELOG entry: Fixed the Activity screen not explaining which MetaMask Pay step failed or why

## **Related issues**

Fixes: [TMCU-1214](https://consensyssoftware.atlassian.net/browse/TMCU-1214)

## **Manual testing steps**

```gherkin
Feature: Failed MetaMask Pay step in Activity details

  Scenario: user sees which step failed
    Given the tmcuActivityRedesignEnabled and tmcuTransactionsRedesignEnabled flags are on
    And the user has a failed Perps or Predict funding row

    When user opens Wallet > Activity and taps that row
    Then the red cross sits on the step the failure explanation names
    And steps after it show as upcoming, with no timestamps

  Scenario: user reads what happened
    When user taps the failed step
    Then a sheet titled "Why it failed" opens
    And it explains where the funds ended up
    And its "View on block explorer" button opens the failed leg's transaction

  Scenario: the sheet does not linger over the webview
    When user taps "View on block explorer" in the sheet
    Then the sheet closes and the block explorer opens

  Scenario: cancelled deposits are explained too
    Given the user cancelled a Pay-funded Perps deposit

    When user taps the crossed step
    Then the sheet explains the cancellation instead of doing nothing
```

Forcing a real leg failure on demand is impractical. To exercise each bucket locally, seed a Pay parent with legs into `TransactionController` state and a bridge-history entry keyed by the leg id — `COMPLETE` with `quote.destAsset.address` matching the surface's destination constant reaches `bridgedNotDeposited`; a failed relay leg reaches `bridgeFailed`.

## **Screenshots/Recordings**

`~`

### **Before**

`~`

### **After**

`~`

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

Labels and the performance checks below are deliberately unchecked: they are outstanding author work, not N/A.

#### Performance checks (if applicable)

- [ ] I've tested on Android
  - Ideally on a mid-range device; emulator is acceptable
- [ ] I've tested with a power user scenario
  - Use these [power-user SRPs](https://consensyssoftware.atlassian.net/wiki/spaces/TL1/pages/edit-v2/401401446401?draftShareId=9d77e1e1-4bdc-4be1-9ebb-ccd916988d93) to import wallets with many accounts and tokens
- [ ] I've instrumented key operations with Sentry traces for production performance metrics
  - See [`trace()`](/app/util/trace.ts) for usage and [`addToken`](/app/components/Views/AddAsset/components/AddCustomToken/AddCustomToken.tsx#L274) for an example

For performance guidelines and tooling, see the [Performance Guide](https://consensyssoftware.atlassian.net/wiki/spaces/TL1/pages/400085549067/Performance+Guide+for+Engineers).

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

[TMCU-1214]: https://consensyssoftware.atlassian.net/browse/TMCU-1214?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---

> [!NOTE]
> **Medium Risk**
> Changes how failed funding rows behave and reads TransactionController state for leg errors and explorer links, but it is read-only presentation with no signing or payment logic.
> 
> **Overview**
> Failed MetaMask Pay funding steps in Activity details no longer look like ordinary timeline dots. **Failed steps use a shared red cross** (`StepFailureIcon`) instead of an error dot, and **tapping a failed step with a message opens a “Why it failed” bottom sheet** rather than jumping straight to the block explorer. The sheet shows the classified error text and a **View on block explorer** button that targets the failed leg when known; the explorer button can run **`onNavigate` first** so the sheet dismisses before the webview opens.
> 
> **`usePayFundsFailure`** (with **`useActivityLocalTransaction`** and **`getTransactionErrorMessage`**) resolves live parent/leg transactions, picks approval vs relay failures earliest-first, and supplies message, `failedLeg`, and per-leg explorer targets. **Perps and Predict deposit** step builders pass that into the timeline so the cross lands on the right step (approval, relay/bridge, or last step by default); withdrawals largely keep explorer-on-row behavior. **`useActivityPayMetadata`** now reads Pay metadata from the same live transaction hook so fees and failure state stay aligned.
> 
> New test IDs, locale strings under `activity_details.failure`, and broad unit tests cover the sheet, timeline interactions, and hook edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c95a18e86569cb896a5248a506a08458c4c2805f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: app/components/Views/confirmations/components/activity/transaction-details-status/transaction-details-status.tsx [hunk 1]
```diff
@@ -50,8 +48,10 @@ export function TransactionDetailsStatus({
 
   const solutionText =
     !text && status === TransactionStatus.failed && hasSuccessfulPerpsBridge
-      ? strings('transaction_details.perps_deposit_solution', {
+      ? strings('activity_details.failure.bridged_not_deposited', {
```
Comment (matthewwalsh0) [inline app/components/Views/confirmations/components/activity/transaction-details-status/transaction-details-status.tsx]: I'm afraid this code path isn't in use anymore as we exclusively use Relay which atomically deposits to Perps.

Meaning any bridge references are all deprecated and pending removal.

File: app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+  bridgeHistory,
+  destination,
+}: {
+  parent: TransactionMeta;
+  legs: TransactionMeta[];
+  legIds: string[];
+  bridgeHistory: Record<string, BridgeHistoryItem>;
+  destination: PayFundsDestination;
+}): { shape: PayFundsFailureShape; leg?: TransactionMeta } {
+  const failedApproval = legs.find(
```
Comment (matthewwalsh0) [inline app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts]: Have all these alternate messages been approved by @bschorchit ?

I'm not sure I see the value in adding this complex "shape" parsing in the UI.

I'd vote to maintain what we currently do in the existing MetaMask Pay activity which is simply to surface the underlying `TransactionMeta` error for any transaction leg.

The nature of those transactions, and the underlying logic is entirely owned by the `TransactionPayController` and associated pay strategy, so it is uniquely placed to provide detailed and useful errors via the standard `error` property from the `TransactionController` transactions.

File: app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+  );
+  if (failedApproval) {
+    return { shape: 'approvalFailed', leg: failedApproval };
+  }
+
+  const failedBridge = legs.find(
+    (leg) => hasTransactionType(leg, RELAY_DEPOSIT_TYPES) && hasFailed(leg),
+  );
+  const offChainFailedLegId = legIds.find(
+    (id) => bridgeHistory[id]?.status?.status === StatusTypes.FAILED,
```
Comment (matthewwalsh0) [inline app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts]: As mentioned in the other file, all references to bridge history can be removed as we no longer support the legacy bridge flow.

File: app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+      return strings('activity_details.failure.bridge_failed');
+    case 'approvalFailed':
+      return strings('activity_details.failure.approval_failed');
+    case 'notSubmitted':
+      return strings('activity_details.failure.not_submitted');
+    case 'cancelled':
+      return strings('activity_details.failure.cancelled');
+    case 'unknown':
+    default:
+      return strings('activity_details.failure.unknown');
```
Comment (matthewwalsh0) [inline app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts]: To clarify, this entire switch should instead just be reading the `TransactionMeta` error whenever available.

My understanding of this current code is we never present the `error` property so we're losing incredibly valuable debug information.

Whether it is developer facing or not is a decision for the controller as it presents the most useful and relevant information it has available.

Your examples in the description are actual on-chain revert messages so actually extremely valuable for isolating and debugging specific errors.

File: app/components/Views/ActivityDetails/templates/PerpsDetails.tsx
Comment (matthewwalsh0) [inline app/components/Views/ActivityDetails/templates/PerpsDetails.tsx:1]: The existing MetaMask Pay activity was architected to be generic, so it could support any new flow without duplication and only minimum additions if custom copy was required.

Crucially, it also ensured a consistent UX for any MetaMask Pay flow rather than divergence and alternate patterns between different features. 

We should consolidate to a single `TransactionPay` template to define a generic and single reference for any MetaMask Pay activity view.

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: app/components/Views/confirmations/components/activity/transaction-details-status/transaction-details-status.tsx (hunk 1 above)
Comment (vinnyhoward) [reply app/components/Views/confirmations/components/activity/transaction-details-status/transaction-details-status.tsx]: Got it, thanks for the callout I will remove

File: app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts (hunk 2 above)
Comment (vinnyhoward) [reply app/components/Views/ActivityDetails/hooks/usePayFundsFailure.ts]: @matthewwalsh0 They were created as placeholders. I wanted to get feedback from your team and perps on how I am deriving these errors before finalizing the copy. Let me check out the `TransactionMeta` errors to see if that is what we want and run it by Thomas
